framework/src/utils/RankTwoTensor.C

```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+                                            std::vector<T> & d,
+                                            T e[3]) const
+{
+  //  This is derived from the Algol procedures tred2 by
+  //  Bowdler, Martin, Reinsch, and Wilkinson, Handbook for
+  //  Auto. Comp., Vol.ii-Linear Algebra, and the corresponding
+  //  Fortran subroutine in EISPACK.
+  int N = 3;
+
+  for (int j = 0; j < N; j++)
+    d[j] = V(N - 1, j);
+
+  // Householder reduction to tridiagonal form.
+  for (int i = N - 1; i > 0; i--)
+  {
+    // Scale to avoid under/overflow.
+    T scale = 0.0;
+    T h = 0.0;
+    for (int k = 0; k < i; k++)
+      scale = scale + std::fabs(d[k]);
+    if (scale == 0.0)
+    {
+      e[i] = d[i - 1];
+      for (int j = 0; j < i; j++)
+      {
+        d[j] = V(i - 1, j);
+        V(i, j) = 0.0;
+        V(j, i) = 0.0;
+      }
+    }
+    else
+    {
+      // Generate Householder vector.
+      for (int k = 0; k < i; k++)
+      {
+        d[k] /= scale;
+        h += d[k] * d[k];
+      }
+      T f = d[i - 1];
+      T g = std::sqrt(h);
+      if (f > 0)
+        g = -g;
+      e[i] = scale * g;
+      h = h - f * g;
+      d[i - 1] = f - g;
+      for (int j = 0; j < i; j++)
+        e[j] = 0.0;
+
+      // Apply similarity transformation to remaining columns.
+      for (int j = 0; j < i; j++)
+      {
+        f = d[j];
+        V(j, i) = f;
+        g = e[j] + V(j, j) * f;
+        for (int k = j + 1; k <= i - 1; k++)
+        {
+          g += V(k, j) * d[k];
+          e[k] += V(k, j) * f;
+        }
+        e[j] = g;
+      }
+      f = 0.0;
+      for (int j = 0; j < i; j++)
+      {
+        e[j] /= h;
+        f += e[j] * d[j];
+      }
+      T hh = f / (h + h);
+      for (int j = 0; j < i; j++)
+        e[j] -= hh * d[j];
+      for (int j = 0; j < i; j++)
+      {
+        f = d[j];
+        g = e[j];
+        for (int k = j; k <= i - 1; k++)
+          V(k, j) -= (f * e[k] + g * d[k]);
+        d[j] = V(i - 1, j);
+        V(i, j) = 0.0;
+      }
+    }
+    d[i] = h;
+  }
+
+  // Accumulate transformations.
+  for (int i = 0; i < N - 1; i++)
+  {
+    V(N - 1, i) = V(i, i);
+    V(i, i) = 1.0;
+    T h = d[i + 1];
+    if (h != 0.0)
+    {
+      for (int k = 0; k <= i; k++)
+        d[k] = V(k, i + 1) / h;
+      for (int j = 0; j <= i; j++)
+      {
+        T g = 0.0;
+        for (int k = 0; k <= i; k++)
+          g += V(k, i + 1) * V(k, j);
+        for (int k = 0; k <= i; k++)
+          V(k, j) -= g * d[k];
+      }
+    }
+    for (int k = 0; k <= i; k++)
+      V(k, i + 1) = 0.0;
+  }
+  for (int j = 0; j < N; j++)
+  {
+    d[j] = V(N - 1, j);
+    V(N - 1, j) = 0.0;
+  }
+  V(N - 1, N - 1) = 1.0;
+  e[0] = 0.0;
+}
+
+template <typename T>
+void
+RankTwoTensorTempl<T>::tridiagonalQLDecomposition(RankTwoTensorTempl<T> & V,
+                                                  std::vector<T> & d,
+                                                  T e[3]) const
```

Comment:
If possible let's avoid raw C arrays. You could pass in a reference to a `std::array<T, RankTwoTensorTempl<T>:: N>`.